core/cmd/local_client.go

```diff
@@ -59,43 +60,44 @@ func openDB(cfg config.GeneralConfig, lggr logger.Logger) (db *sqlx.DB, err erro
 	return
 }
 
-func applicationLockDB(cfg config.GeneralConfig, db *sqlx.DB, lggr logger.Logger, sig shutdown.Signal) (cleanup func(), err error) {
-	lggr.Debugf("Using database locking mode: %s", cfg.DatabaseLockingMode())
-
+func applicationLockDB(ctx context.Context, cfg config.GeneralConfig, db *sqlx.DB, lggr logger.Logger) (func(), error) {
 	lockingMode := cfg.DatabaseLockingMode()
-	cleanup = func() {}
 
-	// Lease will be explicitly released on application stop
+	lggr.Debugf("Using database locking mode: %s", lockingMode)
+
+	// Lease will be explicitly released on ctx cancellation
 	// Take the lease before any other DB operations
 	var leaseLock pg.LeaseLock
 	switch lockingMode {
 	case "lease", "dual":
 		leaseLock = pg.NewLeaseLock(db, cfg.AppID(), lggr, cfg.LeaseLockRefreshInterval(), cfg.LeaseLockDuration())
-		if err = leaseLock.TakeAndHold(sig); err != nil {
-			return cleanup, errors.Wrap(err, "failed to take initial lease on database")
+		if err := leaseLock.TakeAndHold(ctx); err != nil {
+			return nil, errors.Wrap(err, "failed to take initial lease on database")
 		}
-		cleanup = leaseLock.Release
 	}
 
 	// Try to acquire an advisory lock to prevent multiple nodes starting at the same time
+	// The lock will be explicitly released on ctx cancellation
 	var advisoryLock pg.AdvisoryLock
 	switch lockingMode {
 	case "advisorylock", "dual":
 		advisoryLock = pg.NewAdvisoryLock(db, cfg.AdvisoryLockID(), lggr, cfg.AdvisoryLockCheckInterval())
-		if err = advisoryLock.TakeAndHold(); err != nil {
-			return cleanup, errors.Wrap(err, "error acquiring lock")
+		if err := advisoryLock.TakeAndHold(ctx); err != nil {
+			if leaseLock != nil {
+				leaseLock.Release()
+			}
```

Comment:
Why are we releasing the lease lock here?

I suspect there may be some confusion, which is understandable since I haven't explained this very well (or at all).

In `"dual"` mode we use both lease lock AND advisory lock. Lease lock is taken first, then advisory (I don't think it matters which order they are released in).

`"dual"` mode is used in order to keep compatibility with older versions of chainlink. Say we want to move to using lease locking, but we are currently running nodes using the advisory lock. We can use "dual" to respect the advisory lock of the old nodes while also taking a lease lock, allowing us to safely switch to "lease" on the next upgrade. Make sense?

Reply:
This was my suggestion so that the caller doesn't have to handle releasing anymore when an error is returned.

Reply:
Added a possible clarifying comment :shrug:

Reply:
Thank you Jordan!